test/server/cards/02_SHD/upgrades/HeroicResolve.spec.js

```diff
@@ -0,0 +1,170 @@
+                });
+
+                this.p1HeroicResolve = this.player1.findCardByName('heroic-resolve');
+                this.p2HeroicResolve = this.player2.findCardByName('heroic-resolve');
+            });
+
+            it('should work even if it is attached to an opponent\'s card', function () {
+                this.player1.passAction();
+
+                // player 2 attaches their Heroic Resolve to player 1's AT-RT
+                this.player2.clickCard(this.p2HeroicResolve);
+                this.player2.clickCard(this.frontierAtrt);
+
+                this.player1.clickCard(this.frontierAtrt);
+
+                // check exact buttons to make sure that there is only one Heroic Resolve ability prompt
+                expect(this.player1).toHaveExactPromptButtons([
+                    'Attack',
+                    'Attack with this unit. It gains +4/+0 and Overwhelm for this attack.',
+                    'Cancel'
+                ]);
+
+                this.player1.clickPrompt('Attack with this unit. It gains +4/+0 and Overwhelm for this attack.');
+                expect(this.player1).not.toHavePassAbilityButton();
+                expect(this.player1).toBeAbleToSelectExactly([this.p1HeroicResolve, this.p2HeroicResolve]);
+                this.player1.clickCard(this.p1HeroicResolve);
+
+                expect(this.player1).toBeAbleToSelectExactly([this.p2Base, this.wampa, this.specforceSoldier]);
+                this.player1.clickCard(this.wampa);
+                expect(this.frontierAtrt).toHaveExactUpgradeNames(['heroic-resolve']);
+                expect(this.p1HeroicResolve).toBeInLocation('discard');
+                expect(this.wampa).toBeInLocation('discard');
+                expect(this.p2Base.damage).toBe(3); // extra 1 damage from the heroic resolve stat boost
+                expect(this.player1.countExhaustedResources()).toBe(2);
+
+                // activate second Heroic Resolve
+                this.player2.passAction();
+                this.frontierAtrt.exhausted = false;
+                this.frontierAtrt.damage = 0;
+                this.player1.clickCard(this.frontierAtrt);
+                expect(this.player1).toHaveEnabledPromptButton('Attack with this unit. It gains +4/+0 and Overwhelm for this attack.');
+
+                this.player1.clickPrompt('Attack with this unit. It gains +4/+0 and Overwhelm for this attack.');
+                expect(this.player1).not.toHavePassAbilityButton();
+                expect(this.player1).toBeAbleToSelectExactly([this.p2Base, this.specforceSoldier]);
+
+                this.player1.clickCard(this.specforceSoldier);
+                expect(this.frontierAtrt.isUpgraded()).toBe(false);
+                expect(this.p2HeroicResolve).toBeInLocation('discard');
+                expect(this.specforceSoldier).toBeInLocation('discard');
+                expect(this.p2Base.damage).toBe(8);
+                expect(this.player1.countExhaustedResources()).toBe(4);
+            });
```

Comment:
Do you think we might want to make sure that the ability can't defeat a Heroic Resolve on another unit?